monty-31/src/aarch64_neon/packing.rs

```diff
@@ -586,38 +617,261 @@ impl_packed_field_pow_2!(
+#[inline]
+unsafe fn dot_product_4<P, LHS, RHS>(lhs: &[LHS; 4], rhs: &[RHS; 4]) -> PackedMontyField31Neon<P>
+where
+    P: FieldParameters + MontyParametersNeon,
+    LHS: IntoVec<P>,
+    RHS: IntoVec<P>,
+{
+    unsafe {
+        // Accumulate the full 64-bit sum C = Σ lhs_i ⋅ rhs_i.
+        //
+        // Start with lhs_0 ⋅ rhs_0 (split into low lanes 0,1 and high lanes 2,3).
+        let (mut sum_l, mut sum_h) = vmull_wide_u32(lhs[0].into_vec(), rhs[0].into_vec());
+
+        // Add lhs_1 ⋅ rhs_1.
+        sum_l = aarch64::vmlal_u32(
+            sum_l,
+            aarch64::vget_low_u32(lhs[1].into_vec()),
+            aarch64::vget_low_u32(rhs[1].into_vec()),
+        );
+        sum_h = aarch64::vmlal_high_u32(sum_h, lhs[1].into_vec(), rhs[1].into_vec());
+
+        // Add lhs_2 ⋅ rhs_2.
+        sum_l = aarch64::vmlal_u32(
+            sum_l,
+            aarch64::vget_low_u32(lhs[2].into_vec()),
+            aarch64::vget_low_u32(rhs[2].into_vec()),
+        );
+        sum_h = aarch64::vmlal_high_u32(sum_h, lhs[2].into_vec(), rhs[2].into_vec());
+
+        // Add lhs_3 ⋅ rhs_3.
+        sum_l = aarch64::vmlal_u32(
+            sum_l,
+            aarch64::vget_low_u32(lhs[3].into_vec()),
+            aarch64::vget_low_u32(rhs[3].into_vec()),
+        );
+        sum_h = aarch64::vmlal_high_u32(sum_h, lhs[3].into_vec(), rhs[3].into_vec());
+
+        // Split C into 32-bit halves per lane:
+        // - c_lo = C mod 2^{32},
+        // - c_hi = C >> 32.
+        let c_lo = aarch64::vuzp1q_u32(
+            aarch64::vreinterpretq_u32_u64(sum_l),
+            aarch64::vreinterpretq_u32_u64(sum_h),
+        );
+        let c_hi = aarch64::vuzp2q_u32(
+            aarch64::vreinterpretq_u32_u64(sum_l),
+            aarch64::vreinterpretq_u32_u64(sum_h),
+        );
+
+        // Partial high-word reduction: c_hi' ∈ [0,P).
+        //
+        // Since C < 4P^2 and P < 2^{31}, we have c_hi < 2P.
+        //
+        // Map [P,2P) → [0,P) by subtracting P.
+        let c_hi_sub = aarch64::vsubq_u32(c_hi, P::PACKED_P);
+        let c_hi_prime = aarch64::vminq_u32(c_hi, c_hi_sub);
+
+        // q ≡ c_lo ⋅ μ (mod 2^{32}), with μ = −P^{-1} (mod 2^{32}).
+        let q = aarch64::vmulq_s32(aarch64::vreinterpretq_s32_u32(c_lo), P::PACKED_MU);
+        let q_u32 = aarch64::vreinterpretq_u32_s32(q);
+
+        // Compute (q⋅P)_hi = high 32 bits of q⋅P per lane (exact unsigned widening multiply).
+        let (qp_l, qp_h) = vmull_wide_u32(q_u32, P::PACKED_P);
+        let qp_hi = aarch64::vuzp2q_u32(
+            aarch64::vreinterpretq_u32_u64(qp_l),
+            aarch64::vreinterpretq_u32_u64(qp_h),
+        );
+
+        // Approximate reduced value: d ≈ c_hi' − (q⋅P)_hi.
+        //
+        // This ignores a possible borrow from the low word; the error is at most 1.
+        let d = aarch64::vsubq_s32(
+            aarch64::vreinterpretq_s32_u32(c_hi_prime),
+            aarch64::vreinterpretq_s32_u32(qp_hi),
+        );
```

Comment:
No a borrow is impossible. This is one of the points of monty multiplication, mod 2^32 c and qp are equal.
```suggestion
        let d = aarch64::vsubq_s32(
            aarch64::vreinterpretq_s32_u32(c_hi_prime),
            aarch64::vreinterpretq_s32_u32(qp_hi),
        );
```